[MRG][FEATURE] LiDAR monitoring metric for Joint-Embedding Trainer

sami-bg opened the pull request:
## Description

Implements [LiDAR](https://arxiv.org/pdf/2312.04000), as per the suggestion [here](https://github.com/rbalestr-lab/stable-SSL/pull/138#issuecomment-2558647238). Since LiDAR requires multiple views, I had to do some slight refactors to the monitor code. Most notably:
```
    # NOTE From the LiDAR paper:
    # "We adopt the terminology in (Garrido et al., 2022), and refer to the
    # output of the encoder f as a representation, and the output of the composed encoder and projector
    # e = ψ◦f as an embedding"
    #
    # My assumption is that all SSL methods have `representations` but only `JEA` draws the distinction
    # between `representations` and `embeddings`.
```
Because of this, I had to rename some things and add an additional property to the `JointEmbeddingTrainer` that keeps track of the most recent `embeddings`. I still need to think through whether any of this can be parallelized in the DDP setting or if we need to gather all embeddings to the rank0 device before computing LiDAR.

## Checklist

<!-- - Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] I have read the [**Contributing**](https://rbalestr-lab.github.io/stable-SSL.github.io/dev/contributing.html#) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR to the [**RELEASES.rst**](../RELEASES.rst) file.

sami-bg commented:
As per #143 ! Let me know if this looks good :) and happy new year!

huguesva commented on stable_ssl/monitors.py:
```diff
@@ -4,14 +4,53 @@
+    return num_devices
+
+
+def gather_to_rank0(x: torch.Tensor):
```
could you please put these gather functions in utils/utils.py instead ? try to use the functions we already have in utils or modify them to fit your use case, also maybe we could put the "if rank 0" condition in the BaseTrainer when we call the monitor instead of inside this gather function? Idk what's your opinion on this?

huguesva commented on stable_ssl/monitors.py:
```diff
@@ -23,7 +62,7 @@ class Monitor:
     name: str = "monitor"
 
-    def compute(self, x):
+    def compute(self, x: BaseTrainer):
```
not a big fan of the name "x" for a BaseTrainer variable, maybe call it "trainer"?

huguesva commented on stable_ssl/monitors.py:
```diff
@@ -37,49 +76,163 @@ def __init__(self, limit: int = 8, epsilon: float = 1e-7):
+        return self.lidar(embeddings)
+
+
+if __name__ == "__main__":
```
could you remove this part below "if __name__ == "__main__":" and instead try to make a test in the tests folder for the monitor? I know it's not easy to find good tests do eventually we can do it in a separate PR later.

huguesva reviewed (changes requested):
looks very nice ! I left a few comments about unifying "gather" functions and making a clean pytest file in the tests folder (we have to start thinking about test coverage!). 
Thanks Sami!

sami-bg commented:
Good ideas @huguesva ! I'll address them soon

sami-bg commented:
Howdy! A few changes:
1. Both monitoring metrics now broadcast the final value to each rank, so that if you choose to independently log the metric, it will not be stale. This is because currently, only rank0 computes the metrics from the reduced/gathered data.
2. I updated the JointEmbeddingTrainer subclasses to use latest_representation and latest_embedding, but I am still adhering to the note on the semantics of 'representation' and 'embedding' from the LiDAR paper. Because of this, the DINO class has something like `latest_representation = embedding` and `latest_embedding = projections`. I have no strong opinions about this, of course, so feel free to let me know if you want it named differently!
3. I added tests for RankMe and LiDAR. This includes testing it in the DDP case, which emulates multiple devices by using the gloo backend. It uses torch.multiprocessing to spawn multiple processes and assigns each one a rank.
4. All tests pass! But I'm not too sure what the best way to test metrics like RankMe and LiDAR are. I did my best, but ofc feel free to let me know if you have any other unit test ideas.
5. The DDP tests take between 15 and 50 seconds to run. We can also consider skipping them by default, and only run them when the branch is merged to main or something.

huguesva commented:
Thank you so much @sami-bg !! It looks great. 

About the notations, I think you are right I didn't adopt the most popular names. Probably we should change embeddings -> representations and projections -> embedding. Agreed @RandallBalestriero ? If so don't worry I will do the changes. :)

huguesva reviewed (approved):
Looking great!
